woocommerce/woocommerce
# [testing workflows] Add optional jobs in CI workflow

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/45097.

Evaluate optional jobs in `ci.yml`. Optional jobs configuration is already supported since https://github.com/woocommerce/woocommerce/pull/47261. 

Some context on the approach:
GitHub doesn't provide a straightforward way to evaluate the dynamic matrix jobs conclusions (see more info in [45097](https://github.com/woocommerce/woocommerce/issues/45097)). The jobs outputs doesn't support output from each job and will only write the output from the last job from a dynamic group (e.g. `project-e2e-test-jobs`) and we cannot match failed jobs with the optional configuration value coming from the matrix.

The workaround chosen here is to use the Github API to get the failed jobs and to use the name of the jobs to mark optional ones, by appending the `(optional)` string to their name.

I went with a defensive approach, by assuming each job is required unless it's explicitly specified otherwise (by job name or included in the ignore list of the evaluation script). 

There was another workaround, proposed in https://github.com/woocommerce/woocommerce/issues/45097 which implied using artefacts to store the job context and download and evaluate the artefacts. While it should work, I find it too verbose, requiring too many extra steps and adding too many possible failure points. 

The job name when they are skipped have become even uglier and possibly more confusing, like there was an error, e.g. `CI / Lint - ${{ matrix.projectName }} ${{ ((matrix.optional && ' (optional)') || '') }} (pull_request)`. I'm thinking on fixing this on a follow-up.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

You can test the script locally by running with `TEST_MODE` env variable, which will load the jobs data from `evaluate-jobs-conclusions-test-data.json` file instead of fetching from Github API.

Run `TEST_MODE=1 node ./.github/workflows/scripts/evaluate-jobs-conclusions.js`

You can also run it with real data by setting the required `REPOSITORY`, `RUN_ID`, and `GITHUB_TOKEN` env variables. It should work with any existing run, but the problem is that older runs don't have the optional job names and all jobs will be required.

If you want to test it in CI, you could force different jobs to fail as needed.

What's the expected result:
- A required job is any job that doesn't end with `(optional)`.
- A required job that failed or was cancelled should fail the evaluation.
- An optional job that failed or was cancelled should not fail the evaluation.

A successful run with all tests running: [9098126672](https://github.com/woocommerce/woocommerce/actions/runs/9098126672/job/25008777597)
